lib/validator/v3/validator.go

```diff
@@ -695,6 +696,24 @@ func validateRule(v *validator.Validate, structLevel *validator.StructLevel) {
 		}
 	}
 
+	icmp := numorstring.ProtocolFromString("ICMP")
+	icmpv6 := numorstring.ProtocolFromString("ICMP")
+	if rule.ICMP != nil && (rule.Protocol == nil || (*rule.Protocol != icmp && *rule.Protocol != icmpv6)) {
+		structLevel.ReportError(reflect.ValueOf(rule.ICMP), "ICMP", "", reason(protocolIcmpMsg))
+	}
+
+	// Check that the IPVersion of the protocol matches the IPVersion of the ICMP protocol.
+	if (rule.Protocol != nil && *rule.Protocol == icmp) || (rule.NotProtocol != nil && *rule.NotProtocol == icmp) {
+		if rule.IPVersion != nil && *rule.IPVersion != 4 {
```

Comment:
This will let nil through, but I think we should validate that the protocol is not nil.  I.e. `if rule.IPVersion == nil || *rule.IPVersion != 4`

Reply:
@fasaxc this change is causing problems in our downstream components. Any rule that includes ICMP rules now also needs to set IPVersion=4. Did we mean to require that?

If so, i'll probably need to update this comment:
https://github.com/projectcalico/libcalico-go/blob/master/lib/apis/v3/policy.go#L38-L40

Reply:
Looks like Neil found some code I'd forgotten about in Felix's `ipVersionToProtoIPVersion` that does auto-set the version based on the presence of ICMP/ICMPv6 so maybe we _can_ relax this.  Sorry about that, that code wasn't where I expected it in Felix.

It does look like we need to validate against using ICMP with IPv6 and ICMPv6 with IPv4, though.